core/dbt/adapters/base/impl.py

```diff
@@ -1162,6 +1162,47 @@ def get_rows_different_sql(
 
         return sql
 
+    def valid_incremental_strategies(self):
+        return ["append", "delete+insert"]
+
+    @available.parse_none
+    def get_incremental_strategy_macro(self, strategy: str):
+        # Construct macro_name from strategy name
+        if strategy is None:
+            strategy = "default"
+
+        # validate strategies for this adapter
+        valid_strategies = self.valid_incremental_strategies()
+        valid_strategies.append("default")
+        builtin_strategies = ["append", "delete+insert", "merge", "insert_overwrite"]
+        if strategy in builtin_strategies and strategy not in valid_strategies:
+            raise RuntimeException(
+                f"The incremental strategy '{strategy}' is not valid for this adapter"
+            )
+
+        strategy = strategy.replace("+", "_")
+        macro_name = f"get_incremental_{strategy}_sql"
+        # get a macro and return it
+        manifest = self._macro_manifest
+        macro = manifest.find_macro_by_name(macro_name, self.config.project_name, None)
+        if not macro:
+            raise RuntimeException(
+                'dbt could not find an incremental strategy macro with the name "{}" in {}'.format(
+                    macro_name, self.config.project_name
+                )
+            )
+
+        from dbt.context.providers import generate_runtime_macro_context
+
+        macro_context = generate_runtime_macro_context(
+            macro=macro,
+            config=self.config,
```

Comment:
[BLOCKING] Of all my comments, I think this is the only one we _must_ address before merging.

I believe this is causing the test failure in https://github.com/dbt-labs/dbt-snowflake/pull/196, because `config.get('merge_update_columns')` here is returning `None`:

https://github.com/dbt-labs/dbt-core/blob/e9e811eb99e736c7028f4d52db4b9da8c860e17a/core/dbt/include/global_project/macros/materializations/models/incremental/merge.sql#L8

Why? The `config` is different! I think this line, in `generate_runtime_macro_context`, may have the effect of setting a _different_ `config`. I dropped [our Jinja debugger](https://docs.getdbt.com/reference/dbt-jinja-functions/debug-method) into that line above, and saw this:
```
ipdb> l_1_config
<dbt.context.providers.RuntimeConfigObject object at 0x13b10c4f0>
ipdb> l_1_config.get("merge_update_columns")
ipdb> l_1_config.__dict__
{'model': ParsedMacro(unique_id='macro.dbt.get_incremental_default_sql', package_name='dbt', root_path='/Users/jerco/dev/product/dbt-core/core/dbt/include/global_project', path='macros/materializations/models/incremental/strategies.sql', original_file_path='macros/materializations/models/incremental/strategies.sql', name='get_incremental_default_sql', macro_sql="{% macro get_incremental_default_sql(arg_dict) %}\n\n  {{ return(adapter.dispatch('get_incremental_default_sql', 'dbt')(arg_dict)) }}\n\n{% endmacro %}", resource_type=<NodeType.Macro: 'macro'>, tags=[], depends_on=MacroDependsOn(macros=['macro.dbt_snowflake.snowflake__get_incremental_default_sql']), description='', meta={}, docs=Docs(show=True), patch_path=None, arguments=[], created_at=1658322950.229271)}
```

Whereas this is the value of `config` within the incremental materialization:
```
ipdb> l_1_config
<dbt.context.providers.RuntimeConfigObject object at 0x1382a6d30>
ipdb> l_1_config.get("merge_update_columns")
['email', 'ip_address']
ipdb> l_1_config.__dict__
{'model': CompiledModelNode(raw_sql='{{\n    config(\n        materialized = "incremental",\n        unique_key = "id",\n        merge_update_columns = ["email", "ip_address"]\n    )\n}}\n\n\nselect *\nfrom {{ ref(\'seed\') }}\n\n{% if is_incremental() %}\n\n    where load_date > (select max(load_date) from {{this}})\n\n{% endif %}', compiled=True, database='DBT_TEST', schema='TEST16583249731264382551_SIMPLE_COPY', fqn=['test', 'incremental_update_cols'], unique_id='model.test.incremental_update_cols', package_name='test', root_path='/private/var/folders/7h/hj5_fw9j291c58hwfdvy5xbm0000gp/T/dbt-int-test-frmfwd6h', path='incremental_update_cols.sql', original_file_path='models-merge-update/incremental_update_cols.sql', name='incremental_update_cols', resource_type=<NodeType.Model: 'model'>, alias='incremental_update_cols', checksum=FileHash(name='sha256', checksum='e16fc1080522481ea1692289e54b2c30a243084f9e5019e5f6fc05d05c83800b'), config=NodeConfig(_extra={'merge_update_columns': ['email', 'ip_address']}, enabled=True, alias=None, schema=None, database=None, tags=[], meta={}, materialized='incremental', incremental_strategy=None, persist_docs={}, post_hook=[], pre_hook=[], quoting={}, column_types={}, full_refresh=None, unique_key='id', on_schema_change='ignore', grants={}), _event_status={'node_status': <RunningStatus.Executing: 'executing'>}, tags=[], refs=[['seed']], sources=[], metrics=[], depends_on=DependsOn(macros=['macro.dbt.is_incremental', 'macro.dbt_snowflake.set_query_tag', 'macro.dbt.should_full_refresh', 'macro.dbt.load_relation', 'macro.dbt.make_temp_relation', 'macro.dbt.incremental_validate_on_schema_change', 'macro.dbt.run_hooks', 'macro.dbt.create_table_as', 'macro.dbt.run_query', 'macro.dbt.process_schema_changes'], nodes=['seed.test.seed']), description='', columns={}, meta={}, docs=Docs(show=True), patch_path=None, compiled_path='target/compiled/test/models-merge-update/incremental_update_cols.sql', build_path=None, deferred=False, unrendered_config={'materialized': 'incremental', 'unique_key': 'id', 'merge_update_columns': ['email', 'ip_address']}, created_at=1658324991.3003168, config_call_dict={'materialized': 'incremental', 'unique_key': 'id', 'merge_update_columns': ['email', 'ip_address']}, compiled_sql='\n\n\nselect *\nfrom DBT_TEST.TEST16583249731264382551_SIMPLE_COPY.seed\n\n\n\n    where load_date > (select max(load_date) from DBT_TEST.TEST16583249731264382551_SIMPLE_COPY.incremental_update_cols)\n\n', extra_ctes_injected=True, extra_ctes=[], relation_name='DBT_TEST.TEST16583249731264382551_SIMPLE_COPY.incremental_update_cols', _pre_injected_sql=None)}
```

Two options come to mind:
1. Pass the `config` in as a second argument to `adapter.get_incremental_strategy_macro()`, so that it can also be passed in here as well (instead of using `self.config`). Or somehow find a way to use the same model context from the materialization, to resolve these macros. This feels finicky to me, and I wasn't able to get it working after a very quick attempt, but it might be doable.
2. Pass in `merge_update_columns`, and `sql_header`, and any other "grab-from-global-state" configs as a keyword arguments to the macro (`get_merge_sql`). That's actually the pattern we want, ultimately! I'm just not sure if there's a way to do this while preserving backwards compatibility for the macro signature. It also removes the ability for users to define a new config in the model, and _just_ override the one macro they need (`get_merge_sql`), without copy-paste-editing the whole incremental materialization.

Reply:
I'm passing in the context on the call to get_incremental_strategy_macro. The python code can just return the callable macro from the context.

Reply:
@gshank It works!! Nice one :)